numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -158,12 +159,50 @@ raise_casting_error(
         return -1;
     }
 
+    int nin = ufunc->nin;
+    int nop = nin + ufunc->nout;
+    PyObject *froms = PyTuple_New(nop);
+    if (froms == NULL) { 
+        return -1;
+    }
+    PyObject *tos = PyTuple_New(nop);
+    if (tos == NULL) { 
+        return -1;
+    }
+
+    if (input) {
+        for (int j = 0; j < nop; j++) {
+            if (operands[j] == NULL) {
+                PyTuple_SET_ITEM(froms, j, Py_None);
+            }
+            else {
+                Py_INCREF(operands[j]);
+                PyTuple_SET_ITEM(froms, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(tos, j, (PyObject *)dtypes[j]);
+        }
+    }
+    else {
+        for (int j = 0; j < nop; j++) {
+            Py_INCREF(dtypes[j]);
+            PyTuple_SET_ITEM(froms, j, (PyObject *)dtypes[j]);
+            if (operands[j] == NULL) {
+                PyTuple_SET_ITEM(tos, j, Py_None);
+            }
+            else {
+                Py_INCREF(operands[j]);
+                PyTuple_SET_ITEM(tos, j, (PyObject *)PyArray_DESCR(operands[j]));
+            }
+        }
+    }
+
     exc_value = Py_BuildValue(
         "ONOOi",
```

Comment:
```suggestion
        "ONNNi",
```
These are new references, so you either need to decref them after the call, or pass the ownership to the tuple.